File: sources/Hera b1/dev/hera/client/utils/MovementUtils.java
package dev.hera.client.utils;

import net.minecraft.client.Minecraft;

public class MovementUtils {

    private static Minecraft mc = Minecraft.getMinecraft();

    public static float getDirection() {
        float yaw = mc.thePlayer.rotationYaw;
        float forward = 1f;

        if(mc.thePlayer.moveForward < 0f)
            yaw += 180f;

        if(mc.thePlayer.moveForward < 0f)
            forward = -0.5f;
        else if(mc.thePlayer.moveForward > 0f)
            forward = 0.5f;

        if(mc.thePlayer.moveStrafing > 0f)
            yaw -= 90 * forward;
        else if(mc.thePlayer.moveStrafing < 0f)
            yaw += 90f * forward;

        yaw *= 0.017453292F;
        return yaw;

    }

    public static boolean isMoving(){
        return mc.thePlayer != null && (mc.thePlayer.movementInput.moveForward != 0f || mc.thePlayer.movementInput.moveStrafe != 0f);
    }

    public static void strafe(double speed) {
        if(!isMoving())
            return;
        double yaw = getDirection();
        mc.thePlayer.motionX = -Math.sin(yaw) * speed;
        mc.thePlayer.motionZ = Math.cos(yaw) * speed;

    }

}

File: sources/Leaf devEnv/com/leafclient/leaf/utils/_MovementUtils.kt
package com.leafclient.leaf.utils

import com.leafclient.leaf.event.game.entity.PlayerMoveEvent
import net.minecraft.client.Minecraft
import net.minecraft.client.entity.EntityPlayerSP
import net.minecraft.entity.MoverType
import net.minecraft.entity.player.EntityPlayer
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

val EntityPlayerSP.isMoving: Boolean
    get() = movementInput.moveForward != 0.0F || movementInput.moveStrafe != 0.0F

/**
 * Strafes towards the movement direction with a velocity of [velocity]
 */
fun PlayerMoveEvent.strafe(
    velocity: Double = sqrt(x * x + z * z)
) {
    val dir = Minecraft.getMinecraft().player.movementYaw
    x = -sin(dir) * velocity
    z = cos(dir) * velocity
}

fun PlayerMoveEvent.safeWalk(offsetY: Double) {
    val player = Minecraft.getMinecraft().player
    if ((type == MoverType.SELF || type == MoverType.PLAYER) && player.onGround) {
        while (x != 0.0 && player.world
                .getCollisionBoxes(player, player.entityBoundingBox.offset(x, offsetY, 0.0)).isEmpty()) {
            if (x < 0.05 && x >= -0.05) {
                x = 0.0
            } else if (x > 0.0) {
                x -= 0.05
            } else {
                x += 0.05
            }
        }

        while (z != 0.0 && player.world
                .getCollisionBoxes(player, player.entityBoundingBox.offset(0.0, offsetY, z)).isEmpty()
        ) {
            if (z < 0.05 && z >= -0.05) {
                z = 0.0
            } else if (z > 0.0) {
                z -= 0.05
            } else {
                z += 0.05
            }
        }

        while (x != 0.0 && z != 0.0 && player.world
                .getCollisionBoxes(player, player.entityBoundingBox.offset(x, offsetY, z)).isEmpty()
        ) {
            if (x < 0.05 && x >= -0.05) {
                x = 0.0
            } else if (x > 0.0) {
                x -= 0.05
            } else {
                x += 0.05
            }
            if (z < 0.05 && z >= -0.05) {
                z = 0.0
            } else if (z > 0.0) {
                z -= 0.05
            } else {
                z += 0.05
            }
        }
    }
}